Sources/ParseSwift/Coding/ParseEncoder.swift

```diff
@@ -0,0 +1,259 @@
+    }
+}
+
+// MARK: _ParseEncoderKeyedEncodingContainer
+internal struct _ParseEncoderKeyedEncodingContainer<Key: CodingKey>: KeyedEncodingContainerProtocol {
+    let codingPath: [CodingKey]
+    let dictionary: NSMutableDictionary
+    let skippedKeys: Set<String>
+
+    init(codingPath: [CodingKey], dictionary: NSMutableDictionary, skippingKeys: Set<String>) {
+        self.codingPath = codingPath
+        self.dictionary = dictionary
+        self.skippedKeys = skippingKeys
+    }
+
+    mutating func encodeNil(forKey key: Key) throws {
+        if skippedKeys.contains(key.stringValue) { return }
+
+        dictionary[key.stringValue] = nil
+    }
+
+    mutating func encode<T>(_ value: T, forKey key: Key) throws where T: Encodable {
+        if skippedKeys.contains(key.stringValue) { return }
+
+        dictionary[key.stringValue] = try _ParseEncoder.encode(
+            value,
+            with: codingPath + [key],
+            skippingKeys: skippedKeys
+        )
+    }
+
+    mutating func nestedContainer<NestedKey>(
+        keyedBy keyType: NestedKey.Type,
+        forKey key: Key
+    ) -> KeyedEncodingContainer<NestedKey> where NestedKey: CodingKey {
+        let container = _ParseEncoderKeyedEncodingContainer<NestedKey>(
+            codingPath: codingPath + [key],
+            dictionary: dictionary,
+            skippingKeys: skippedKeys
+        )
+
+        return KeyedEncodingContainer(container)
+    }
+
+    mutating func nestedUnkeyedContainer(forKey key: Key) -> UnkeyedEncodingContainer {
+        _ParseEncoderUnkeyedEncodingContainer(
+            codingPath: codingPath + [key],
+            dictionary: dictionary,
+            skippingKeys: skippedKeys
+        )
+    }
+
+    mutating func superEncoder() -> Encoder {
+        fatalError()
+    }
+
+    mutating func superEncoder(forKey key: Key) -> Encoder {
+        fatalError()
+    }
+}
+
+// MARK: _ParseEncoderSingleValueEncodingContainer
+internal struct _ParseEncoderSingleValueEncodingContainer: SingleValueEncodingContainer {
+    let codingPath: [CodingKey]
+    let dictionary: NSMutableDictionary
+    let skippedKeys: Set<String>
+
+    init(codingPath: [CodingKey], dictionary: NSMutableDictionary, skippingKeys: Set<String>) {
+        self.codingPath = codingPath
+        self.dictionary = dictionary
+        self.skippedKeys = skippingKeys
+    }
+
+    var key: String {
+        codingPath.last?.stringValue ?? "<root>"
+    }
+
+    mutating func encodeNil() throws {
+        dictionary[key] = nil
+    }
+
+    mutating func encode<T>(_ value: T) throws where T: Encodable {
+        dictionary[key] = try _ParseEncoder.encode(value, with: codingPath, skippingKeys: skippedKeys)
+    }
+}
+
+// MARK: _ParseEncoderUnkeyedEncodingContainer
+internal struct _ParseEncoderUnkeyedEncodingContainer: UnkeyedEncodingContainer {
+    let codingPath: [CodingKey]
+    let dictionary: NSMutableDictionary
+    let skippedKeys: Set<String>
+
+    var array: NSMutableArray {
```

Comment:
same here with NS, also the forced casting doesn't seem like the best approach. If we can't cast, we should be throwing, errorring with feedback and not crashing, or worse case fatalError

Reply:
We can't `throw` in a computed property, but I'll add a `fatalError` with a message.

I'd say if this code path ever executes, it's a serious problem because that means someone else tried to encode the same thing into the same dictionary with a different type, *after* this unkeyed coding container was called to do the same thing, and that's definitely a crash-worthy error.